_extension/src/util.ts

```diff
@@ -0,0 +1,91 @@
+import * as path from "path";
+import * as vscode from "vscode";
+import packageJson from "../package.json";
+
+const version = packageJson.version;
+
+export const jsTsLanguageModes = [
+    "typescript",
+    "typescriptreact",
+    "javascript",
+    "javascriptreact",
+];
+
+export const builtinTSExtensionId = "vscode.typescript-language-features";
+
+export interface ExeInfo {
+    path: string;
+    version: string;
+}
+
+export function getBuiltinExePath(context: vscode.ExtensionContext): string {
+    return context.asAbsolutePath(path.join("./lib", `tsgo${process.platform === "win32" ? ".exe" : ""}`));
+}
+
+function workspaceResolve(relativePath: string): vscode.Uri {
+    if (path.isAbsolute(relativePath)) {
+        return vscode.Uri.file(relativePath);
+    }
+    if (vscode.workspace.workspaceFolders && vscode.workspace.workspaceFolders.length > 0) {
+        const workspaceFolder = vscode.workspace.workspaceFolders[0];
+        return vscode.Uri.joinPath(workspaceFolder.uri, relativePath);
+    }
+    return vscode.Uri.file(relativePath);
+}
+
+export async function getExe(context: vscode.ExtensionContext): Promise<ExeInfo> {
+    const config = vscode.workspace.getConfiguration("typescript.native-preview");
+    const exeName = `tsgo${process.platform === "win32" ? ".exe" : ""}`;
+
+    let exe = config.get<string>("tsdk");
+    if (exe) {
+        if (exe.endsWith("/@typescript/native-preview")) {
+            try {
+                const packagePath = workspaceResolve(exe);
+                const packageJsonPath = vscode.Uri.joinPath(packagePath, "package.json");
+                const packageJson = JSON.parse(await vscode.workspace.fs.readFile(packageJsonPath).then(buffer => buffer.toString()));
+                const getExePath = (await import(vscode.Uri.joinPath(packagePath, "lib", "getExePath.js").toString())).default;
```

Comment:
Does this require modifying the package to be importable? IIRC this isn't exported on the package. Does this only work because the emitted code is CJS?

We should probably just formally export it (and maybe drop the default export?).

Reply:
No, this works because it’s an absolute path, so the package.json is irrelevant.

Reply:
Also relative. Only bare specifiers do the complicated thing. The idea being that ESM resolution should generally be predictable and cheap, but package specifier resolution is what it is.